[ADF-5142] move js-api ACS wrapper from ACA

adinapitul opened the pull request:
- move js-api ACS wrapper from ACA
- change share-file.e2e.ts to use the new apis as an example

**Please check if the PR fulfills these requirements**

> - [x] The commit message follows our [guidelines](https://github.com/Alfresco/alfresco-ng2-components/wiki/Commit-format)
> - [x] Tests for the changes have been added (for bug fixes / features)
> - [ ] Docs have been added / updated (for bug fixes / features)

<!--
 Before submitting your PR, please check that your code follows our contribution guidelines:
 https://github.com/Alfresco/alfresco-ng2-components/wiki/Code-contribution-acceptance-criteria
 -->

**What kind of change does this PR introduce?** (check one with "x")

> - [ ] Bugfix
> - [ ] Feature
> - [ ] Code style update (formatting, local variables)
> - [ ] Refactoring (no functional changes, no api changes)
> - [ ] Build related changes
> - [ ] Documentation
> - [ ] Other... Please describe:


**What is the current behaviour?** (You can also link to an open issue here)
- at the moment, the testing package contains just a few methods for interacting with the ACS apis
- ACA has this wrapper over the js-api ACS calls, containing lots of useful methods very much needed in the e2e tests for creating preconditions
- these methods are now needed in ADW for the extensions e2e tests


**What is the new behaviour?**
- having this wrapper in the testing package will allow any adf app to easily create the needed preconditions for the e2e tests, without having to reinvent the wheel
- in addition, a nice part of this wrapper is that we would no longer have to login before every js-api request. It's all being handled like this:

```
const adminAcsApi = new RepoClient(); // default is using the admin credentials
const userAcsApi = new RepoClient(acsUser.id, acsUser.password); // this instance uses the user credentials

await userAcsApi.nodes.createFile(...); // createFile is performed as the user
await adminAcsApi.nodes.createFile(...); // createFile is performed as admin
```


**Does this PR introduce a breaking change?** (check one with "x")

> - [ ] Yes
> - [ ] No
> - [x] Not yet

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

mauriziovitale commented on lib/testing/src/lib/content-services/rest-apis/upload/upload-api.ts:
```diff
@@ -0,0 +1,68 @@
+    super(username, password);
+  }
+
+//   async uploadFile(fileName: string, parentFolderId: string = '-my-') {
```
do we need this part?

adinapitul replied:
yes, it's needed as well, I had it commented out cause I needed to talk with @rdiacenco  and @cristinaj  about how to proceed with this one, since this has parts that overlap with already existing upload.actions.
we will keep both in parallel for now, until we can change the test files to use this new upload-api, then we can remove upload.actions.

adinapitul replied:
my next commit will add the upload-api as well

mauriziovitale commented on lib/testing/src/lib/content-services/rest-apis/repo-client.ts:
```diff
@@ -0,0 +1,85 @@
+        return new QueriesApi(this.auth.username, this.auth.password);
+    }
+
+    // get upload() {
```
do we need it ?

eromano commented on lib/testing/src/lib/content-services/rest-apis/favorites/favorites-api.ts:
```diff
@@ -0,0 +1,166 @@
+      };
+      return await ApiUtil.retryCall(favorite);
+    } catch (error) {
+      // this.handleError(`${this.constructor.name} ${this.isFavoriteWithRetry.name}`, error);
```
commented code

eromano commented on lib/testing/src/lib/content-services/rest-apis/public-api.ts:
```diff
@@ -0,0 +1,31 @@
+export * from './queries/queries-api';
+export * from './trashcan/trashcan-api';
+export * from './search/search-api';
+// export * from './upload/upload-api';
```
commented code can you remove it?

eromano commented on lib/testing/src/lib/content-services/rest-apis/people/people-api-models.ts:
```diff
@@ -0,0 +1,45 @@
+ * limitations under the License.
+ */
+
+export interface PersonModel {
```
exists a similar model in the js-api can we use it?

adinapitul replied:
this model is similar, but still different than what it is in js-api. I tried to use that one, but I don't get the same flexibility as with this model built strictly for the e2e simple needs (for example, some specific fields are required in Person from js-api, but in the e2e I don't want to enter those info every single time, and this model helps with simplicity: all you need to pass is the username)

eromano commented on lib/testing/src/lib/content-services/rest-apis/nodes/node-content-tree.ts:
```diff
@@ -0,0 +1,79 @@
+    description?: string;
+}
+
+export function flattenNodeContentTree(content: NodeContentTree, relativePath: string = '/'): NodeBodyCreate[] {
```
not sure if this is the right place to export flattenNodeContentTree

adinapitul replied:
do you have another suggestion where should this be placed and exported?

eromano commented on lib/testing/src/lib/content-services/rest-apis/nodes/node-body-create.ts:
```diff
@@ -0,0 +1,31 @@
+export const NODE_TITLE = 'cm:title';
+export const NODE_DESCRIPTION = 'cm:description';
+
+export class NodeBodyCreate {
```
NodeBodyCreate exist already in the JS-API

adinapitul replied:
they are similar, but still have differences. This one I think it's more simple, fit for what the e2e need

eromano commented on lib/testing/src/lib/content-services/rest-apis/authentication/authentication-api.ts:
```diff
@@ -0,0 +1,34 @@
+
+    async logout() {
+      try {
+        await this.apiLogin();
```
why do we login before?

eromano commented on lib/testing/src/lib/content-services/rest-apis/repo-client.ts:
```diff
@@ -0,0 +1,79 @@
+import { AuthenticationApi } from './authentication/authentication-api';
+
+export class RepoClient {
+    constructor(
```
this constructor has the wrong indentation. Can you create the JS-API in the constructor here otherwise all the time that you do the new in the single API you create a new instance

eromano commented on lib/testing/src/lib/content-services/rest-apis/repo-client.ts:
```diff
@@ -0,0 +1,79 @@
+import { UploadApi } from './upload/upload-api';
+import { AuthenticationApi } from './authentication/authentication-api';
+
+export class RepoClient {
```
Can we name it ContentApi? it matches the js-api modules

eromano commented on lib/testing/src/lib/content-services/rest-apis/upload/upload-api.ts:
```diff
@@ -0,0 +1,69 @@
+
+import * as fs from 'fs';
+
+export class UploadApi extends RepoApi {
```
we already have UploadActions this seems a duplication

adinapitul replied:
upload-api and upload.actions indeed overlap a bit. Talking with Cristina and Roxana, and from what I saw in another PR, it's best we go with these 2 in parallel for now, and create tasks for changing the test files to move from the old upload.actions to the new upload-api
upload.actions also contains methods belonging to nodes-api, those need to be replaced with new calls and then removed.

adinapitul commented:
> There are some returns null can you remove it?

can you please be more specific? there are methods that return null in the catch, if these are the ones you are referring to, but without that return, I'm getting the error that not all code paths return a value

> Some methods are missing the return type can we add it?

Done.

> I see different any as type can we try to use the right type?

can you please be more specific? I think most of, if not all the methods return the same thing as the called method from js-api

> can you remove the login all the time we use a method and make the instantiation only in the RepoClient?

I tried that, but didn't get to a nice and clean solution.
The way this is built now, you need to instantiate only ContentClient, and the rest of the apis are getters. I didn't find a way to have an async getter. 
In any case, even if it looks like the login is done multiple times, it basically works like a check-if-I'm-logged-in, cause the first call creates the ticket, the next calls will get the same ticket if one already exists and is valid. This is how I noticed this endpoint works like.

adinapitul commented:
@eromano , I did a commit where I removed all the login calls and moved it in a setup method, but this was causing some tests to fail. Here's an example of what happens:
```create instance of user API: like: userApi = await new ContentApi(username, password).setup();```
```
1. user API call: create file1
2. UI: login with the same user
3. user API call: create file2
4. UI: logout
5. user API call: create file3
```
Step 5 fails. It seems that when you logout from the UI, the ticket for that user is destroyed, so the user API calls that follow are no longer authenticated.

So, with this in mind, I think we should just let the login calls where they are, considering what I mentioned previously, that these calls do not mean new logins are done every time, but just a check that the user ticket is valid.

eromano commented:
I will set up a meeting about this PR, For the following problems:

- In testing, there are already 2 different ways of setup the services and this PR introduce a third
-In testing, there are already 3 ways of calling the services and I would like to know what we should keep
- In testing, we are replicating a lot of service business logic that is already present in the core library
